### PR TITLE
Add Abi Encode support

### PR DESCRIPTION
**Pull request by rodrigo-pino**
Can now transpile:
- abi.encode
- abi.encodePacked
- abi.encodeWithSelector
- abi.encodeWithSignature

### CODE REVIEW

**Review by GuyMcComb — COMMENTED** [review]
Could I be a pain and ask for a test that combines dynamic types and inline types?
Such as abi.encode(uint[], unit8, uint8[]), does this increase the complexity a lot?

**Review by RohitRanjangit — CHANGES_REQUESTED** [review]
PR with thunders ⚡️ !!

can we have one more case in behaviour tests which involves Enums ? 

PS: Will  provide more reviews , some time later

File: src/utils/nodeTypeProcessing.ts [hunk 1]
```diff
@@ -263,3 +263,112 @@ export function isStorageSpecificType(
...
+  }
+  if (type instanceof FixedBytesType) {
+    return type.size;
+  }
+  if (type instanceof AddressType) {
+    return 20;
+  }
+  if (
+    type instanceof BoolType ||
+    (type instanceof UserDefinedType && type.definition instanceof EnumDefinition)
```
[the remark this answers is not in the document]
Comment (rodrigo-pino) [reply src/utils/nodeTypeProcessing.ts:290]: Enums cannot have more than 256 elements, so they are similar to a uint8 and that's why it's byte size is 1. Doc for reference:
https://docs.soliditylang.org/en/v0.8.16/types.html?highlight=enums#enums

**Review by RohitRanjangit — APPROVED** [review]

**Comment by GuyMcComb** [comment]
Sem Tests Passed.

File: src/cairoUtilFuncGen/abi/base.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import { isValueType } from '../../utils/nodeTypeProcessing';
+import { typeNameFromTypeNode } from '../../utils/utils';
+import { StringIndexedFuncGenWithAuxiliar } from '../base';
+
+export abstract class AbiBase extends StringIndexedFuncGenWithAuxiliar {
+  protected functionName = 'not_implemented';
+
+  public gen(expressions: Expression[], sourceUnit?: SourceUnit): FunctionCall {
+    const exprTypes = expressions.map(
+      (expr) => generalizeType(getNodeType(expr, this.ast.compilerVersion))[0],
```
Comment (cicr99) [inline src/cairoUtilFuncGen/abi/base.ts:21]: Maybe `safeGetNodeType` should be used here instead of `getNodeType`

**Review by cicr99 — APPROVED** [review]

**Pull request merged**